ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java

```diff
@@ -56,6 +61,9 @@ public abstract class LogAppenderBase implements LogAppender {
   private final AwaitForSignal eventAwaitForSignal;
 
   private final AtomicBoolean heartbeatTrigger = new AtomicBoolean();
+  private final List<BiFunction<Boolean, AppendEntriesReplyProto, Boolean>> heartbeatAckListeners;
```

Comment:
It only needs a single listener queue in the ReadRequests, instead of a queue per LogAppender.  Each listener in the queue corresponds to a readIndex.  When a LogAppender receive an appendEntries reply, just notify ReadRequests so that it can update the listeners.

Reply:
done